client/azure_client_set.go

```diff
@@ -11,12 +11,194 @@ import (
+			Objectives: map[float64]float64{0.5: 0.05, 0.9: 0.01, 0.99: 0.001},
+		},
+		[]string{"status_code"},
+	)
+
+	storageAccountsClientMetricsDesc = prometheus.NewSummaryVec(
+		prometheus.SummaryOpts{
+			Name:       "azure_sdk_client_storageaccounts",
+			Help:       "HTTP metrics for Azure SDK Storage Accounts client.",
+			Objectives: map[float64]float64{0.5: 0.05, 0.9: 0.01, 0.99: 0.001},
+		},
+		[]string{"status_code"},
+	)
+
+	usageClientMetricsDesc = prometheus.NewSummaryVec(
+		prometheus.SummaryOpts{
+			Name:       "azure_sdk_client_usage",
+			Help:       "HTTP metrics for Azure SDK Usage client.",
+			Objectives: map[float64]float64{0.5: 0.05, 0.9: 0.01, 0.99: 0.001},
+		},
+		[]string{"status_code"},
+	)
+
+	virtualNetworkMetricsDesc = prometheus.NewSummaryVec(
+		prometheus.SummaryOpts{
+			Name:       "azure_sdk_client_virtualnetwork",
+			Help:       "HTTP metrics for Azure SDK Virtual Network client.",
+			Objectives: map[float64]float64{0.5: 0.05, 0.9: 0.01, 0.99: 0.001},
+		},
+		[]string{"status_code"},
+	)
+
+	virtualNetworkGatewaysMetricsDesc = prometheus.NewSummaryVec(
+		prometheus.SummaryOpts{
+			Name:       "azure_sdk_client_virtualnetworkgateways",
+			Help:       "HTTP metrics for Azure SDK Virtual Network Gateways client.",
+			Objectives: map[float64]float64{0.5: 0.05, 0.9: 0.01, 0.99: 0.001},
+		},
+		[]string{"status_code"},
+	)
+
+	virtualNetworkGatewayConnectionsMetricsDesc = prometheus.NewSummaryVec(
+		prometheus.SummaryOpts{
+			Name:       "azure_sdk_client_virtualnetworkgatewayconnections",
+			Help:       "HTTP metrics for Azure SDK Virtual Network Gateway Connections client.",
+			Objectives: map[float64]float64{0.5: 0.05, 0.9: 0.01, 0.99: 0.001},
+		},
+		[]string{"status_code"},
+	)
+
+	virtualMachineScaleSetsMetricsDesc = prometheus.NewSummaryVec(
+		prometheus.SummaryOpts{
+			Name:       "azure_sdk_client_virtualmachinescalesets",
+			Help:       "HTTP metrics for Azure SDK Virtual Machine ScaleSets client.",
+			Objectives: map[float64]float64{0.5: 0.05, 0.9: 0.01, 0.99: 0.001},
+		},
+		[]string{"status_code"},
+	)
+
+	virtualMachineScaleSetVMsMetricsDesc = prometheus.NewSummaryVec(
+		prometheus.SummaryOpts{
+			Name:       "azure_sdk_client_virtualmachinescalesetvms",
+			Help:       "HTTP metrics for Azure SDK Virtual Machine ScaleSet VMs client.",
+			Objectives: map[float64]float64{0.5: 0.05, 0.9: 0.01, 0.99: 0.001},
+		},
+		[]string{"status_code"},
+	)
+
+	vnetPeeringMetricsDesc = prometheus.NewSummaryVec(
+		prometheus.SummaryOpts{
+			Name:       "azure_sdk_client_vnetpeering",
+			Help:       "HTTP metrics for Azure SDK VNET Peering client.",
+			Objectives: map[float64]float64{0.5: 0.05, 0.9: 0.01, 0.99: 0.001},
+		},
+		[]string{"status_code"},
+	)
+)
+
+func init() {
+	// Register prometheus metrics.
+	err := prometheus.Register(deploymentsClientMetricsDesc)
+	if err != nil && !collector.IsAlreadyRegisteredError(err) {
```

Comment:
`IsAlreadyRegisteredError()` is needed for unit tests.

Reply:
I thought this is what we have the collector set thingy for. So that you do not have to do the crappy init magic. Why doing in that way here?

Reply:
why don't we register the metrics in the `MetricsCollector` factory method so that they only get registered once?

Reply:
> I thought this is what we have the collector set thingy for. So that you do not have to do the crappy init magic. Why doing in that way here?

`client` package is used from `service/collector` so we can't import anything from there (import loop).

> why don't we register the metrics in the MetricsCollector factory method so that they only get registered once?

Because currently `MetricsCollector` gets called on every reconciliation loop because clients are constructed in `initCtx`.

Reply:
Sounds like a use case to extend `exporterkit` maybe.

Reply:
You can define descs in collector package, init then there and import them here.